### PR TITLE
Add metrics to express data cache

### PR DESCRIPTION
**Pull request by muddyfish**
## Description of change

Adds metrics to express data cache
Fixes a bug where getting a cache miss would be reported as an error rather than a cache miss

Relevant issues: N/A

## Does this change impact existing behavior?

Adds metrics, no user facing functionality changes.

## Does this change need a changelog entry in any of the crates?

No

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3-client/src/s3_crt_client/get_object.rs [hunk 1]
```diff
@@ -114,11 +116,14 @@ impl S3CrtClient {
                 let _ = sender.unbounded_send(Ok((offset, data.into())));
             },
             move |result| {
-                // FIXME - Ideally we'd include a reason why we failed here.
-                object_headers_setter_on_finish.or_set(Err(ObjectHeadersError::UnknownError));
                 if result.is_err() {
-                    Err(parse_get_object_error(result).map(ObjectClientError::ServiceError))
+                    let err = parse_get_object_error(result);
```
Comment (passaro) [inline mountpoint-s3-client/src/s3_crt_client/get_object.rs:120]: This is problematic: the result has not been processed yet - see [these](https://github.com/awslabs/mountpoint-s3/blob/b61f4b966f616ed3d231857403946149520aad2a/mountpoint-s3-client/src/s3_crt_client.rs#L514) [comments](https://github.com/awslabs/mountpoint-s3/blob/b61f4b966f616ed3d231857403946149520aad2a/mountpoint-s3-client/src/s3_crt_client.rs#L648).

I think this is the same mistake that causes issues for PUT: https://github.com/awslabs/mountpoint-s3/issues/1007.

Unless we review the callbacks, the correct thing to do would be to wait for the request to finish and return the fully parsed error.

File: mountpoint-s3/src/data_cache/express_data_cache.rs [hunk 2]
```diff
@@ -99,6 +99,30 @@ where
...
+                Ok(None)
+            }
+            e => {
+                self.emit_failure_metric("get_object_failure", MetricsType::Read);
+                Err(DataCacheError::IoFailure(e.into()))
+            }
+        }
+    }
+
+    fn emit_failure_metric(&self, reason: &'static str, type_: MetricsType) {
```
Comment (passaro) [inline mountpoint-s3/src/data_cache/express_data_cache.rs:120]: I'd suggest inlining the metrics calls. This function makes it less clear what you are tracking.

File: mountpoint-s3/src/data_cache/express_data_cache.rs [hunk 3]
```diff
@@ -113,11 +137,15 @@ where
         block_offset: u64,
         object_size: usize,
     ) -> DataCacheResult<Option<ChecksummedBytes>> {
```
Comment (passaro) [inline mountpoint-s3/src/data_cache/express_data_cache.rs:139]: Too many points of failure.
Shall we factor out the logic into an inner method and only handle the metrics here?

File: mountpoint-s3/src/data_cache/express_data_cache.rs [hunk 4]
```diff
@@ -132,24 +160,31 @@ where
             .await
         {
             Ok(result) => result,
-            Err(ObjectClientError::ServiceError(GetObjectError::NoSuchKey)) => return Ok(None),
-            Err(e) => return Err(e.into()),
+            Err(err) => return self.handle_get_object_err(err),
```
Comment (vladem) [inline mountpoint-s3/src/data_cache/express_data_cache.rs:163]: it looks that `ObjectClientError::ServiceError(GetObjectError::NoSuchKey)` is not even possible here, but I'd check what this [method](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3-client/src/s3_crt_client/get_object.rs#L125) returns

File: mountpoint-s3/src/data_cache/express_data_cache.rs (hunk 3 above)
Comment (muddyfish) [reply mountpoint-s3/src/data_cache/express_data_cache.rs:113]: I'm not planning on refactoring into an inner method, as that makes the reasons harder to propagate

File: mountpoint-s3/src/data_cache/express_data_cache.rs [hunk 5]
```diff
@@ -214,6 +256,17 @@ where
...
 }
 
+#[inline]
+fn emit_failure_metric_read(reason: &'static str) {
+    metrics::counter!("express_data_cache.block_hit").increment(0);
+    metrics::counter!("express_data_cache.block_err", "reason" => reason, "type" => "read").increment(1);
+}
+
+#[inline]
+fn emit_failure_metric_write(reason: &'static str) {
```
Comment (passaro) [inline mountpoint-s3/src/data_cache/express_data_cache.rs:266]: nit, non-blocking: we could just inline

File: mountpoint-s3/src/data_cache/express_data_cache.rs (hunk 3 above)
Comment (passaro) [reply mountpoint-s3/src/data_cache/express_data_cache.rs:113]: An alternative could be to still split into an inner function which records the failure (and reason), but move the hit/miss metric to the wrapper.

Happy to review the approach later, though.

**Review by passaro — APPROVED** [review]
Left a few comments, but completely optional.

**Pull request merged**